File: Data Structures & Algorithms/search-in-rotated-sorted-array-ii/submission-0.java
class Solution {
    public boolean search(int[] arr, int target) {
        int pivot = pivot_finding(arr);
        boolean res = (binarySearch(arr,0,pivot-1,target)||binarySearch(arr,pivot,arr.length-1,target));

        return res;
    }


    public int pivot_finding(int arr[]){
        int l = 0;
        int h = arr.length-1;
        while(l<h){
            int mid = l+(h-l)/2;
            if(arr[mid]>arr[h]){
                l = mid+1;
            }
            else if(arr[mid]<arr[h]){
                h = mid;
            }
            else{
                h--;
            }
        }

        return l;
    }

    public boolean binarySearch(int arr[],int low ,int high,int target){
        int l = low;
        int h = high;
        while(l<=h){
            int mid = l+(h-l)/2;
            if(arr[mid]==target){
                return true;
            }
            else if(arr[mid]>target){
                h = mid-1;
            }
            else{
                l = mid+1;
            }
        }

        return false;
    }
}

File: Data Structures & Algorithms/special-array-i/submission-0.py
class Solution:
    def isArraySpecial(self, arr: List[int]) -> bool:
        if len(arr)==1:
            return True
        
        for i in range(1,len(arr)):
            if arr[i]%2==0 and arr[i-1]%2==1:
                continue
            elif arr[i]%2==1 and arr[i-1]%2==0:
                continue
            else:
                return False
        
        return True
